src/solvers/prop/bdd_expr.cpp

```diff
@@ -109,10 +109,12 @@ void bdd_exprt::from_expr(const exprt &expr)
 ///        i-th variable
 /// \param complement: whether we want the negation of the expression
 ///        represented by r
+/// \param cache: map of already computed values
 static exprt as_expr(
   const bdd_nodet &r,
   const std::vector<exprt> &var_to_expr,
-  bool complement)
+  bool complement,
+  std::unordered_map<long, exprt> &cache)
```

Comment:
Why is `long` the right data type? I think I never got an answer on the other BDD PR why the `id()` function has that type. I am still concerned about it.

Reply:
The right type for Cudd would be something like `DdNode*`, and for other implementations of BDDs this type does not exists. Using `long` is a way of making the type independent of the implementation, though having to cast a pointer to `long` is not nice. Any better idea?

Reply:
Can we use a typedef (as a member of the `bdd_nodet` classes)? Then you would just use `bdd_nodet::id_type` (or whatever the name). With the current set-up we do not require that `bdd_nodet::id()` returns that same type in all implementations, just some type that is usable, e.g., as a key in a map.

(Also, `long` would not be sufficient to store a pointer on 64-bit Windows.)

Reply:
good id